gate-web/src/main/groovy/com/netflix/spinnaker/gate/security/oauth2/SpinnakerUserInfoTokenServices.groovy

```diff
@@ -0,0 +1,128 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.gate.security.oauth2
+
+import com.netflix.spinnaker.gate.security.rolesprovider.UserRolesProvider
+import com.netflix.spinnaker.gate.services.CredentialsService
+import com.netflix.spinnaker.security.User
+import com.squareup.okhttp.HttpUrl
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.boot.context.properties.ConfigurationProperties
+import org.springframework.cloud.security.oauth2.resource.ResourceServerProperties
+import org.springframework.cloud.security.oauth2.resource.UserInfoTokenServices
+import org.springframework.security.authentication.BadCredentialsException
+import org.springframework.security.core.AuthenticationException
+import org.springframework.security.oauth2.common.OAuth2AccessToken
+import org.springframework.security.oauth2.common.exceptions.InvalidTokenException
+import org.springframework.security.oauth2.provider.OAuth2Authentication
+import org.springframework.security.oauth2.provider.OAuth2Request
+import org.springframework.security.oauth2.provider.token.ResourceServerTokenServices
+import org.springframework.security.web.authentication.preauth.PreAuthenticatedAuthenticationToken
+import org.springframework.stereotype.Component
+
+/**
+ * ResourceServerTokenServices is an interface used to manage access tokens. The UserInfoTokenService object is an
+ * implementation of that interface that uses an access token to get the logged in user's data (such as email or
+ * profile). We want to customize the Authentication object that is returned to include our custom (Kork) User.
+ */
+class SpinnakerUserInfoTokenServices implements ResourceServerTokenServices {
+  @Autowired
+  ResourceServerProperties sso
+
+  @Autowired
+  UserInfoTokenServices userInfoTokenServices
+
+  @Autowired
+  CredentialsService credentialsService
+
+  @Autowired
+  UserRolesProvider userRolesProvider
+
+  @Autowired
+  UserInfoMapping userInfoMapping
+
+  @Value('${spring.oauth2.client.userAuthorizationUri}')
+  URL userAuthorizationUri
+
+  @Override
+  OAuth2Authentication loadAuthentication(String accessToken) throws AuthenticationException, InvalidTokenException {
+    OAuth2Authentication oAuth2Authentication = userInfoTokenServices.loadAuthentication(accessToken)
+
+    Map details = oAuth2Authentication.userAuthentication.details as Map
+
+    if (!isUserFromRestrictedDomain(details)) {
```

Comment:
Rather than this being negated, I'd go for `isUserDomainAllowed` or something similar.